contracts/pallet-ibc/src/ics20/mod.rs

```diff
@@ -561,11 +561,187 @@ pub fn full_ibc_denom(packet: &Packet, mut token: PrefixedCoin) -> String {
+struct MemoForward {
+	receiver: String,
+	port: String,
+	channel: String,
+	timeout: String,
+	retries: u64,
+	next: Option<Box<MemoForward>>,
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Debug)]
+struct MemoData {
+	forward: MemoForward,
+}
+
+impl<T> HandleMemo<T> for IbcModule<T>
+where
+	T: Config + Send + Sync + pallet_timestamp::Config,
+	u32: From<<T as frame_system::Config>::BlockNumber>,
+	AccountId32: From<<T as frame_system::Config>::AccountId>,
+{
+	fn execute_memo(
+		packet: &Packet,
+		packet_data: &PacketData,
+		receiver: T::AccountId,
+	) -> Result<(), Ics20Error> {
+		//Handle only memo with IBC forward.
+		//TODO XCM memo unwrap. Need to add logic to handle MEMO with xcm instrucntion as well.
+
+		if packet_data.memo.is_empty() {
+			return Ok(())
+		}
+		let memo: MemoForward = serde_json::from_str(&packet_data.memo).map_err(|_| {
+			Self::emit_memo_execution_failed_event(receiver.clone(), packet_data.memo.clone(), 0);
+			Ics20Error::implementation_specific(format!(
+				"Failed to parse memo : {:?} ",
+				packet_data.memo
+			))
+		})?;
+
+		let prefixed_coin = if is_receiver_chain_source(
+			packet.source_port.clone(),
+			packet.source_channel,
+			&packet_data.token.denom,
+		) {
+			let prefix = TracePrefix::new(packet.source_port.clone(), packet.source_channel);
+			let mut c = packet_data.token.clone();
+			c.denom.remove_trace_prefix(&prefix);
+			c
+		} else {
+			let prefix =
+				TracePrefix::new(packet.destination_port.clone(), packet.destination_channel);
+			let mut c = packet_data.token.clone();
+			c.denom.add_trace_prefix(prefix);
+			c
+		};
+
+		// At this point the asset SHOULD exist
+		let asset_id =
+			<T as crate::Config>::IbcDenomToAssetIdConversion::from_denom_to_asset_id(
+				&prefixed_coin.denom.to_string(),
+			)
+			.map_err(|_| {
+				log::warn!(target: "pallet_ibc", "Asset does not exist for denom: {}", prefixed_coin.denom.to_string());
+				Self::emit_memo_execution_failed_event(receiver.clone(), packet_data.memo.clone(), 1);
+				Ics20Error::implementation_specific("asset does not exist".to_string())
+			})?;
+
+		let amount = packet_data.token.amount.as_u256().low_u128();
+
+		let raw_bytes = memo.receiver.as_bytes().to_vec();
+
+		let account_id =
+			crate::MultiAddress::<<T as frame_system::Config>::AccountId>::Raw(raw_bytes.clone());
+		let origin = RawOrigin::Signed(receiver.clone());
+		let channel_id = memo.channel.parse().map_err(|_| {
+			Self::emit_memo_execution_failed_event(receiver.clone(), packet_data.memo.clone(), 2);
+			Ics20Error::implementation_specific("Failed to parse channel ID".to_string())
+		})?;
+
+		//timestamp that current timestamp + 30 minutes
+		let timestamp = <pallet_timestamp::Pallet<T> as frame_support::traits::UnixTime>::now()
+			.as_secs() + 1800;
+
+		let params = crate::TransferParams::<<T as frame_system::Config>::AccountId> {
+			to: account_id,
+			source_channel: channel_id, /* TODO source? should it be destination channel id from
```

Comment:
Yes, in memo you put source channel of the intermediate chain

Reply:
Done